File: objective-c/ArcGISSample/Sensor/SensorViewController.m
//
//  SensorViewController.m
//  ArcGISSample
//
//  Created by esrij on 2015/07/27.
//  Copyright (c) 2015年 esrij. All rights reserved.
//

#import "SensorViewController.h"


@implementation SensorViewController



- (void)viewDidLoad
{
    [super viewDidLoad];
    
    
	self.agsMapView = [[AGSMapView alloc] initWithFrame:self.view.bounds];
	[self.view addSubview:self.agsMapView];
    
    //タイルマップサービスレイヤーの追加
    NSURL *url = [NSURL URLWithString:@"http://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer"];
    AGSTiledMapServiceLayer *tiledMapServiceLayer = [[AGSTiledMapServiceLayer alloc] initWithURL:url];
    [self.agsMapView addMapLayer:tiledMapServiceLayer withName:@"Tiled Layer"];
    
    //グラフィックスレイヤーの追加
    AGSGraphicsLayer *graphicsLayer = [AGSGraphicsLayer graphicsLayer];
    [self.agsMapView addMapLayer:graphicsLayer withName:@"Graphics Layer"];
    
    
    self.locationManager = [[CLLocationManager alloc] init];
    self.locationManager.delegate = self;
    [self.locationManager requestWhenInUseAuthorization];

    [self.locationManager startUpdatingLocation];
    [self.locationManager startUpdatingHeading];
    
}


- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations {
    
    CLLocation *location = [locations lastObject];

    //CLLocationManagerで取得した現在位置からポイントを作成
    AGSGeometryEngine *agsGeomEngine = [AGSGeometryEngine defaultGeometryEngine];
    AGSPoint *agsPoint = [AGSPoint pointWithX:location.coordinate.longitude y:location.coordinate.latitude spatialReference:[AGSSpatialReference spatialReferenceWithWKID:4326]];
    AGSPoint *agsProjectedPoint = (AGSPoint *)[agsGeomEngine projectGeometry:agsPoint toSpatialReference: [AGSSpatialReference spatialReferenceWithWKID:102100]];
    
    //ポイントをグラフィックスレイヤーに追加
    AGSGraphicsLayer *graphicsLayer = (AGSGraphicsLayer *)[self.agsMapView mapLayerForName:@"Graphics Layer"];
    AGSSimpleMarkerSymbol *markerSymbol = [AGSSimpleMarkerSymbol simpleMarkerSymbolWithColor:[UIColor blueColor]];
    AGSGraphic *graphic = [AGSGraphic graphicWithGeometry:agsProjectedPoint symbol:markerSymbol attributes:nil];
    [graphicsLayer removeAllGraphics];
    [graphicsLayer addGraphic:graphic];
    
    [self.agsMapView zoomToScale:100000 withCenterPoint:agsProjectedPoint animated:YES];
    
}



- (void)locationManager:(CLLocationManager *)manager didUpdateHeading:(CLHeading *)newHeading {
    
    if (newHeading.headingAccuracy > 0){
        
        //CLLocationManagerで取得した方位に応じてマップを回転
        [self.agsMapView setRotationAngle:newHeading.magneticHeading];
        
    }
    
}


-(void)viewDidDisappear:(BOOL)animated{
    [super viewDidDisappear:animated];

    [self.locationManager stopUpdatingLocation];
    [self.locationManager stopUpdatingHeading];
    
}


- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: swift/ArcGISSample/SensorViewController.swift
//
//  SensorViewController.swift
//  ArcGISSample
//
//  Created by esrij on 2015/07/27.
//  Copyright (c) 2015年 esrij. All rights reserved.
//

import UIKit
import ArcGIS

class SensorViewController: UIViewController, CLLocationManagerDelegate {
    
    
    var agsMapView: AGSMapView!
    var locationManager:CLLocationManager!
    
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        agsMapView = AGSMapView(frame: view.bounds)
        view.addSubview(agsMapView)
        
        //タイルマップサービスレイヤーの追加
        let url = URL(string: "https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer")
        let tiledLyr = AGSTiledMapServiceLayer(url:url)
        agsMapView.addMapLayer(tiledLyr, withName:"Tiled Layer")
        
        //グラフィックスレイヤーの追加
        let agsGraphicsLayer = AGSGraphicsLayer()
        agsMapView.addMapLayer(agsGraphicsLayer, withName:"Graphics Layer")
        
        
        locationManager = CLLocationManager()
        locationManager.delegate = self
        locationManager.requestWhenInUseAuthorization()
        
        locationManager.startUpdatingLocation()
        locationManager.startUpdatingHeading()
        
        
    }
    
    
    
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {

        let location = locations.last
        
        //CLLocationManagerで取得した現在位置からポイントを作成
        let agsGeomEngine = AGSGeometryEngine.default()
        let agsPoint = AGSPoint(x: location!.coordinate.longitude, y: location!.coordinate.latitude, spatialReference: AGSSpatialReference(wkid:4326))
        let agsProjectedPoint = agsGeomEngine?.projectGeometry(agsPoint, to: AGSSpatialReference(wkid:102100)) as! AGSPoint

        //ポイントをグラフィックスレイヤーに追加
        let graphicsLayer = agsMapView.mapLayer(forName: "Graphics Layer") as! AGSGraphicsLayer
        let markerSymbol = AGSSimpleMarkerSymbol(color: UIColor .blue)
        let graphic = AGSGraphic.graphic(with: agsProjectedPoint, symbol: markerSymbol, attributes: nil) as! AGSGraphic
        graphicsLayer.removeAllGraphics()
        graphicsLayer.addGraphic(graphic)
        agsMapView.zoom(toScale: 100000, withCenter: agsProjectedPoint, animated: true)
    }
    
    
    func locationManager(_ manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {
        
        if (newHeading.headingAccuracy > 0){
            
            //CLLocationManagerで取得した方位に応じてマップを回転
            agsMapView.setRotationAngle(newHeading.magneticHeading, animated: true)
            
        }
    }

    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
}
